kv/kvserver: TestRangeTransferLeaseExpirationBased failed

[(kv/kvserver).TestRangeTransferLeaseExpirationBased failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1879773&tab=buildLog) on [master@de48c6d13431061a6227697a917e574464b80983](https://github.com/cockroachdb/cockroach/commits/de48c6d13431061a6227697a917e574464b80983):

```
=== RUN   TestRangeTransferLeaseExpirationBased
--- FAIL: TestRangeTransferLeaseExpirationBased (5.39s)
W200416 06:58:52.483671 582344 kv/kvserver/store.go:1202  [drain] unable to drain cleanly within 5s, service might briefly deteriorate: operation "wait for raft leadership transfer" timed out after 5s: waiting for 1 replicas to transfer their lease away
(1) operation "wait for raft leadership transfer" timed out after 5s
Wraps: (2) secondary error attachment
  | context deadline exceeded
  | (1) context deadline exceeded
  | Error types: (1) context.deadlineExceededError
Wraps: (3) attached stack trace
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining.func3
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1186
  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1169
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestRangeTransferLeaseExpirationBased.func4.2
  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:892
  | runtime.goexit
  | 	/usr/local/go/src/runtime/asm_amd64.s:1357
Wraps: (4) 2 safe details enclosed
Wraps: (5) waiting for 1 replicas to transfer their lease away
Error types: (1) *contextutil.TimeoutError (2) *secondary.withSecondaryError (3) *withstack.withStack (4) *safedetails.withSafeDetails (5) *errors.errorString

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestRangeTransferLeaseExpirationBased PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRangeTransferLeaseExpirationBased.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestRangeTransferLeaseExpirationBased failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1932130&tab=buildLog) on [master@3441dbae616998147b733c9e3f3b42ae3d7c5bd6](https://github.com/cockroachdb/cockroach/commits/3441dbae616998147b733c9e3f3b42ae3d7c5bd6):

```
=== RUN   TestRangeTransferLeaseExpirationBased
--- FAIL: TestRangeTransferLeaseExpirationBased (5.74s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestRangeTransferLeaseExpirationBased PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRangeTransferLeaseExpirationBased.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

I hit this when trying to bors #48666 yesterday.
```
--- FAIL: TestRangeTransferLeaseExpirationBased (5.42s)
    --- PASS: TestRangeTransferLeaseExpirationBased/Transfer (0.09s)
    --- PASS: TestRangeTransferLeaseExpirationBased/TransferWithExtension (0.12s)
    --- PASS: TestRangeTransferLeaseExpirationBased/DrainTransfer (0.10s)
    --- PASS: TestRangeTransferLeaseExpirationBased/DrainTransferWithExtension (0.10s)
    leaktest.go:121: Leaked goroutine: goroutine 488864 [select]:
        github.com/cockroachdb/cockroach/pkg/util/retry.(*Retry).Next(0xc001377de8, 0xc002500b40)
                /go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:116 +0x185
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining.func3(0x48e0440, 0xc002500b40, 0x12a05f200, 0x48e0440)
                /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1179 +0x265
        github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout(0x48e0440, 0xc002500b40, 0x3fe828c, 0x21, 0x12a05f200, 0xc000b8df40, 0x0, 0x0)
                /go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135 +0xbf
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining(0xc0027c3500, 0x1, 0x0)
                /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1169 +0x229
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestRangeTransferLeaseExpirationBased.func4.2(0xc000f7e990)
                /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:892 +0x54
        created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestRangeTransferLeaseExpirationBased.func4
                /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:891 +0x22e
        Leaked goroutine: goroutine 488876 [runnable]:
        context.WithDeadline.func2()
                /usr/local/go/src/context/context.go:412
        created by time.goFunc
                /usr/local/go/src/time/sleep.go:168 +0x44
I200511 02:47:16.971498 488898 vendor/github.com/cockroachdb/pebble/version_set.go:140  [JOB 1] MANIFEST created 000001
I200511 02:47:16.971602 488898 vendor/github.com/cockroachdb/pebble/open.go:270  [JOB 1] WAL created 000002
W200511 02:47:16.978302 488864 kv/kvserver/store.go:1202  [drain] unable to drain cleanly within 5s, service might briefly deteriorate: operation "wait for raft leadership transfer" timed out after 5s: waiting for 1 replicas to transfer their lease away
(1) operation "wait for raft leadership transfer" timed out after 5s
Wraps: (2) secondary error attachment
  | context deadline exceeded
  | (1) context deadline exceeded
  | Error types: (1) context.deadlineExceededError
Wraps: (3) attached stack trace
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining.func3
  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1186
  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
  |     /go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SetDraining
  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1169
  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestRangeTransferLeaseExpirationBased.func4.2
  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_replica_test.go:892
  | runtime.goexit
  |     /usr/local/go/src/runtime/asm_amd64.s:1357
Wraps: (4) 2 safe details enclosed
Wraps: (5) waiting for 1 replicas to transfer their lease away
Error types: (1) *contextutil.TimeoutError (2) *secondary.withSecondaryError (3) *withstack.withStack (4) *safedetails.withSafeDetails (5) *errors.errorString
```
Given the timing of when we first saw this, it seems possible that this was caused by 078249c52c91182de18a63b50a2622956c35ac2b. Specifically, that commit added in this extra step to draining: https://github.com/cockroachdb/cockroach/commit/078249c52c91182de18a63b50a2622956c35ac2b#diff-fec0a771fc07eba1eb7aaeca1cf61c3eR1150.

I can't easily reproduce this under stress. @knz would you like to take a look at this?

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -722,185 +722,191 @@ func (l *leaseTransferTest) ensureLeaderAndRaftState(
 	})
 }
 
-func TestRangeTransferLeaseExpirationBased(t *testing.T) {
+func TestLeaseExpirationBasedRangeTransfer(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	t.Run("Transfer", func(t *testing.T) {
-		l := setupLeaseTransferTest(t)
-		defer l.mtc.Stop()
-		origLease, _ := l.replica0.GetLease()
-		{
-			// Transferring the lease to ourself should be a no-op.
-			if err := l.replica0.AdminTransferLease(context.Background(), l.replica0Desc.StoreID); err != nil {
-				t.Fatal(err)
-			}
-			newLease, _ := l.replica0.GetLease()
-			if !origLease.Equivalent(newLease) {
-				t.Fatalf("original lease %v and new lease %v not equivalent", origLease, newLease)
-			}
-		}
-
-		{
-			// An invalid target should result in an error.
-			const expected = "unable to find store .* in range"
-			if err := l.replica0.AdminTransferLease(context.Background(), 1000); !testutils.IsError(err, expected) {
-				t.Fatalf("expected %s, but found %v", expected, err)
-			}
-		}
-
-		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
+	l := setupLeaseTransferTest(t)
+	defer l.mtc.Stop()
+	origLease, _ := l.replica0.GetLease()
+	{
+		// Transferring the lease to ourself should be a no-op.
+		if err := l.replica0.AdminTransferLease(context.Background(), l.replica0Desc.StoreID); err != nil {
 			t.Fatal(err)
 		}
-
-		// Check that replica0 doesn't serve reads any more.
-		pErr := l.sendRead(0)
-		nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
-		if !ok {
-			t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
+		newLease, _ := l.replica0.GetLease()
+		if !origLease.Equivalent(newLease) {
+			t.Fatalf("original lease %v and new lease %v not equivalent", origLease, newLease)
 		}
-		if !nlhe.LeaseHolder.Equal(&l.replica1Desc) {
-			t.Fatalf("expected lease holder %+v, got %+v",
-				l.replica1Desc, nlhe.LeaseHolder)
+	}
+
+	{
+		// An invalid target should result in an error.
+		const expected = "unable to find store .* in range"
+		if err := l.replica0.AdminTransferLease(context.Background(), 1000); !testutils.IsError(err, expected) {
+			t.Fatalf("expected %s, but found %v", expected, err)
 		}
+	}
 
-		// Check that replica1 now has the lease.
-		l.checkHasLease(t, 1)
+	if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
+		t.Fatal(err)
+	}
 
-		replica1Lease, _ := l.replica1.GetLease()
+	// Check that replica0 doesn't serve reads any more.
+	pErr := l.sendRead(0)
+	nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
+	if !ok {
+		t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
+	}
+	if !nlhe.LeaseHolder.Equal(&l.replica1Desc) {
+		t.Fatalf("expected lease holder %+v, got %+v",
+			l.replica1Desc, nlhe.LeaseHolder)
+	}
 
-		// We'd like to verify the timestamp cache's low water mark, but this is
-		// impossible to determine precisely in all cases because it may have
-		// been subsumed by future tscache accesses. So instead of checking the
-		// low water mark, we make sure that the high water mark is equal to or
-		// greater than the new lease start time, which is less than the
-		// previous lease's expiration time.
-		if highWater := l.replica1.GetTSCacheHighWater(); highWater.Less(replica1Lease.Start) {
-			t.Fatalf("expected timestamp cache high water %s, but found %s",
-				replica1Lease.Start, highWater)
+	// Check that replica1 now has the lease.
+	l.checkHasLease(t, 1)
+
+	replica1Lease, _ := l.replica1.GetLease()
+
+	// We'd like to verify the timestamp cache's low water mark, but this is
+	// impossible to determine precisely in all cases because it may have
+	// been subsumed by future tscache accesses. So instead of checking the
+	// low water mark, we make sure that the high water mark is equal to or
+	// greater than the new lease start time, which is less than the
+	// previous lease's expiration time.
+	if highWater := l.replica1.GetTSCacheHighWater(); highWater.Less(replica1Lease.Start) {
+		t.Fatalf("expected timestamp cache high water %s, but found %s",
+			replica1Lease.Start, highWater)
+	}
+
+}
+
+// TestLeaseExpirationBasedRangeTransferWithExtension make replica1
+// extend its lease and transfer the lease immediately after
+// that. Test that the transfer still happens (it'll wait until the
+// extension is done).
+func TestLeaseExpirationBasedRangeTransferWithExtension(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	l := setupLeaseTransferTest(t)
+	defer l.mtc.Stop()
+	// Ensure that replica1 has the lease.
+	if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
+		t.Fatal(err)
+	}
+	l.checkHasLease(t, 1)
+
+	extensionSem := make(chan struct{})
+	l.setFilter(true, extensionSem)
+
+	// Initiate an extension.
+	renewalErrCh := make(chan error)
+	go func() {
+		lease, _ := l.replica1.GetLease()
+		renewalErrCh <- l.forceLeaseExtension(1, lease)
+	}()
+
+	// Wait for extension to be blocked.
+	<-extensionSem
+	l.waitForTransferBlocked.Store(true)
+	// Initiate a transfer.
+	transferErrCh := make(chan error)
+	go func() {
+		// Transfer back from replica1 to replica0.
+		err := l.replica1.AdminTransferLease(context.Background(), l.replica0Desc.StoreID)
+		// Ignore not leaseholder errors which can arise due to re-proposals.
+		if errors.HasType(err, (*roachpb.NotLeaseHolderError)(nil)) {
+			err = nil
 		}
-	})
+		transferErrCh <- err
+	}()
+	// Wait for the transfer to be blocked by the extension.
+	<-l.transferBlocked
+	// Now unblock the extension.
+	extensionSem <- struct{}{}
+	l.checkHasLease(t, 0)
+	l.setFilter(false, nil)
 
-	// Make replica1 extend its lease and transfer the lease immediately after
-	// that. Test that the transfer still happens (it'll wait until the extension
-	// is done).
-	t.Run("TransferWithExtension", func(t *testing.T) {
-		l := setupLeaseTransferTest(t)
-		defer l.mtc.Stop()
-		// Ensure that replica1 has the lease.
-		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
-			t.Fatal(err)
-		}
-		l.checkHasLease(t, 1)
+	if err := <-renewalErrCh; err != nil {
+		t.Errorf("unexpected error from lease renewal: %+v", err)
+	}
+	if err := <-transferErrCh; err != nil {
+		t.Errorf("unexpected error from lease transfer: %+v", err)
+	}
+}
 
-		extensionSem := make(chan struct{})
-		l.setFilter(true, extensionSem)
+// TestLeaseExpirationBasedDrainTransfer verifies that a draining store attempts to transfer away
+// range leases owned by its replicas.
+func TestLeaseExpirationBasedDrainTransfer(t *testing.T) {
+	defer leaktest.AfterTest(t)()
 
-		// Initiate an extension.
-		renewalErrCh := make(chan error)
-		go func() {
-			lease, _ := l.replica1.GetLease()
-			renewalErrCh <- l.forceLeaseExtension(1, lease)
-		}()
+	l := setupLeaseTransferTest(t)
+	defer l.mtc.Stop()
+	// We have to ensure that replica0 is the raft leader and that replica1 has
+	// caught up to replica0 as draining code doesn't transfer leases to
+	// behind replicas.
+	l.ensureLeaderAndRaftState(t, l.replica0, l.replica1Desc)
+	l.mtc.stores[0].SetDraining(true, nil /* reporter */)
 
-		// Wait for extension to be blocked.
-		<-extensionSem
-		l.waitForTransferBlocked.Store(true)
-		// Initiate a transfer.
-		transferErrCh := make(chan error)
-		go func() {
-			// Transfer back from replica1 to replica0.
-			err := l.replica1.AdminTransferLease(context.Background(), l.replica0Desc.StoreID)
-			// Ignore not leaseholder errors which can arise due to re-proposals.
-			if errors.HasType(err, (*roachpb.NotLeaseHolderError)(nil)) {
-				err = nil
-			}
-			transferErrCh <- err
-		}()
-		// Wait for the transfer to be blocked by the extension.
-		<-l.transferBlocked
-		// Now unblock the extension.
-		extensionSem <- struct{}{}
-		l.checkHasLease(t, 0)
-		l.setFilter(false, nil)
+	// Check that replica0 doesn't serve reads any more.
+	pErr := l.sendRead(0)
+	nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
+	if !ok {
+		t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
+	}
+	if nlhe.LeaseHolder == nil || !nlhe.LeaseHolder.Equal(&l.replica1Desc) {
+		t.Fatalf("expected lease holder %+v, got %+v",
+			l.replica1Desc, nlhe.LeaseHolder)
+	}
 
-		if err := <-renewalErrCh; err != nil {
-			t.Errorf("unexpected error from lease renewal: %+v", err)
-		}
-		if err := <-transferErrCh; err != nil {
-			t.Errorf("unexpected error from lease transfer: %+v", err)
-		}
-	})
+	// Check that replica1 now has the lease.
+	l.checkHasLease(t, 1)
 
-	// DrainTransfer verifies that a draining store attempts to transfer away
-	// range leases owned by its replicas.
-	t.Run("DrainTransfer", func(t *testing.T) {
-		l := setupLeaseTransferTest(t)
-		defer l.mtc.Stop()
-		// We have to ensure that replica0 is the raft leader and that replica1 has
-		// caught up to replica0 as draining code doesn't transfer leases to
-		// behind replicas.
-		l.ensureLeaderAndRaftState(t, l.replica0, l.replica1Desc)
-		l.mtc.stores[0].SetDraining(true, nil /* reporter */)
+	l.mtc.stores[0].SetDraining(false, nil /* reporter */)
+}
 
-		// Check that replica0 doesn't serve reads any more.
-		pErr := l.sendRead(0)
-		nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
-		if !ok {
-			t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
-		}
-		if nlhe.LeaseHolder == nil || !nlhe.LeaseHolder.Equal(&l.replica1Desc) {
-			t.Fatalf("expected lease holder %+v, got %+v",
-				l.replica1Desc, nlhe.LeaseHolder)
-		}
+// TestLeaseExpirationBasedDrainTransferWithExtension verifies that
+// a draining store waits for any in-progress lease requests to
+// complete before transferring away the new lease.
+func TestLeaseExpirationBasedDrainTransferWithExtension(t *testing.T) {
+	defer leaktest.AfterTest(t)()
 
-		// Check that replica1 now has the lease.
-		l.checkHasLease(t, 1)
+	l := setupLeaseTransferTest(t)
+	defer l.mtc.Stop()
+	// Ensure that replica1 has the lease.
+	if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
+		t.Fatal(err)
+	}
+	l.checkHasLease(t, 1)
 
-		l.mtc.stores[0].SetDraining(false, nil /* reporter */)
-	})
+	extensionSem := make(chan struct{})
+	l.setFilter(true, extensionSem)
 
-	// DrainTransferWithExtension verifies that a draining store waits for any
-	// in-progress lease requests to complete before transferring away the new
-	// lease.
-	t.Run("DrainTransferWithExtension", func(t *testing.T) {
-		l := setupLeaseTransferTest(t)
-		defer l.mtc.Stop()
-		// Ensure that replica1 has the lease.
-		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
-			t.Fatal(err)
-		}
-		l.checkHasLease(t, 1)
+	// Initiate an extension.
+	renewalErrCh := make(chan error)
+	go func() {
+		lease, _ := l.replica1.GetLease()
+		renewalErrCh <- l.forceLeaseExtension(1, lease)
+	}()
 
-		extensionSem := make(chan struct{})
-		l.setFilter(true, extensionSem)
+	// Wait for extension to be blocked.
+	<-extensionSem
 
-		// Initiate an extension.
-		renewalErrCh := make(chan error)
-		go func() {
-			lease, _ := l.replica1.GetLease()
-			renewalErrCh <- l.forceLeaseExtension(1, lease)
-		}()
+	// Make sure that replica 0 is up to date enough to receive the lease.
+	l.ensureLeaderAndRaftState(t, l.replica1, l.replica0Desc)
 
-		// Wait for extension to be blocked.
-		<-extensionSem
+	// Drain node 1 with an extension in progress.
+	go func() {
+		l.mtc.stores[1].SetDraining(true, nil /* reporter */)
+	}()
+	// Now unblock the extension.
+	extensionSem <- struct{}{}
 
-		// Make sure that replica 0 is up to date enough to receive the lease.
-		l.ensureLeaderAndRaftState(t, l.replica1, l.replica0Desc)
+	l.checkHasLease(t, 0)
+	l.setFilter(false, nil)
 
-		// Drain node 1 with an extension in progress.
-		go func() {
-			l.mtc.stores[1].SetDraining(true, nil /* reporter */)
-		}()
-		// Now unblock the extension.
-		extensionSem <- struct{}{}
-
-		l.checkHasLease(t, 0)
-		l.setFilter(false, nil)
-
-		if err := <-renewalErrCh; err != nil {
-			t.Errorf("unexpected error from lease renewal: %+v", err)
-		}
-	})
+	if err := <-renewalErrCh; err != nil {
+		t.Errorf("unexpected error from lease renewal: %+v", err)
+	}
 }
 
 // TestRangeLimitTxnMaxTimestamp verifies that on lease transfer, the

--- a/pkg/kv/kvserver/store.go
+++ b/pkg/kv/kvserver/store.go
@@ -1007,9 +1007,30 @@ func (s *Store) SetDraining(drain bool, reporter func(int, string)) {
 		return
 	}
 
+	baseCtx := logtags.AddTag(context.Background(), "drain", nil)
+
+	// In a running server, the code below (transferAllAway and the loop
+	// that calls it) does not need to be conditional on messaging by
+	// the Stopper. This is because the top level Server calls SetDrain
+	// upon a graceful shutdown, and waits until the SetDrain calls
+	// completes, at which point the work has terminated on its own. If
+	// the top-level server is forcefully shut down, it does not matter
+	// if some of the code below is still running.
+	//
+	// However, the situation is different in unit tests where we also
+	// assert there are no leaking goroutines when a test terminates.
+	// If a test terminates with a timed out lease transfer, it's
+	// possible for the transferAllAway() closure to be still running
+	// when the closer shuts down the test server.
+	//
+	// To prevent this, we add this code here which adds the missing
+	// cancel + wait in the particular case where the stopper is
+	// completing a shutdown while a graceful SetDrain is still ongoing.
+	ctx, cancelFn := s.stopper.WithCancelOnStop(baseCtx)
+	defer cancelFn()
+
 	var wg sync.WaitGroup
 
-	ctx := logtags.AddTag(context.Background(), "drain", nil)
 	transferAllAway := func(transferCtx context.Context) int {
 		// Limit the number of concurrent lease transfers.
 		const leaseTransferConcurrency = 100

